src/filters/ES6Ctors.ml

```diff
@@ -0,0 +1,225 @@
+*)
+let rewrite_ctors com =
+	(* we mark classes that need changing by storing them in these two maps *)
+	let needs_ctor_skipping, does_ctor_skipping, inject_super =
+		let l = List.length com.types in
+		Hashtbl.create l, Hashtbl.create l, Hashtbl.create l
+	in
+
+	(*
+		we're using a reference to the root of the inheritance chain so we can easily
+		generate RootClass._hx_skip_constructor expressions
+	*)
+	let rec mark_needs_ctor_skipping cl root_ref p = 
+		(* for non haxe-generated extern classes we can't generate any valid code, so just fail *)
+		if cl.cl_extern && not (Meta.has Meta.HxGen cl.cl_meta) then begin
+			abort "Must call `super()` constructor before accessing `this` in classes derived from an extern class with constructor" p;
+		end;
+		Hashtbl.replace needs_ctor_skipping cl.cl_path root_ref;
+		match cl.cl_super with
+		| Some ({ cl_constructor = Some _ } as cl_super,_) -> mark_needs_ctor_skipping cl_super root_ref p
+		| _ -> root_ref := cl;
+	in
+
+	let e_empty_super_call = (* super() *)
+		let e_super = mk (TConst TSuper) t_dynamic null_pos in
+		mk (TCall (e_super,[])) com.basic.tvoid null_pos
+	in
+
+	let activated = ref false in
+	let mark t =
+		match t with
+		| TClassDecl ({ cl_constructor = Some { cf_expr = Some { eexpr = TFunction tf } }; cl_super = Some (cl_super,_) } as cl) ->
+			if Type.has_constructor cl_super then begin
+				(* if parent class has a constructor, check for `this` accesses before calling `super()` *)
+				let this_before_super = has_this_before_super tf.tf_expr in
+				Option.may (fun e_this_access ->
+					activated := true;
+					let root_ref = ref cl_super in
+					Hashtbl.replace does_ctor_skipping cl.cl_path root_ref;
+					mark_needs_ctor_skipping cl_super root_ref e_this_access.epos;
+				) this_before_super
+			end else begin
+				(* if there was no ctor in the parent class, we still gotta call `super` *)
+				Hashtbl.add inject_super cl.cl_path cl; 
+			end
+		| _ -> ()
+	in
+	List.iter mark com.types;
+
+	if !activated then begin
+		(* just some helper common exprs *)
+		let e_false = (make_bool com.basic false null_pos) in 
+		let e_true = (make_bool com.basic true null_pos) in
+		let e_hx_ctor = (* this._hx_constructor *)
+			let ethis = mk (TConst TThis) t_dynamic null_pos  in
+			mk (TField (ethis, FDynamic ctor_method_name)) t_dynamic null_pos
+		in
+
+		let change t =
+			match t with
+			| TClassDecl ({ cl_constructor = Some ({ cf_expr = Some ({ eexpr = TFunction tf_ctor } as ctor_expr) } as cf_ctor) } as cl) ->
+				let does_ctor_skipping = try Some (Hashtbl.find does_ctor_skipping cl.cl_path) with Not_found -> None in
+
+				let add_hx_ctor_method () =
+					let cf_fun_ctor = mk_field ctor_method_name cf_ctor.cf_type cf_ctor.cf_pos null_pos in
+					cf_fun_ctor.cf_expr <- Some (replace_super_call ctor_expr);
+					cf_fun_ctor.cf_kind <- Method MethNormal;
+					cl.cl_ordered_fields <- cf_fun_ctor :: cl.cl_ordered_fields;
```

Comment:
@Simn I wonder if adding items to `tclass.cl_ordered_fields` and `tclass.cl_fields` in generator will mess up the compilation server?

Reply:
It should be fine because `cl_restore` will restore it on the next compilation.

Reply:
Yeah it should be fine, because at least Java/C# does that A LOT :)